### PR TITLE
Fix local_map deferred inlining to functionalize mutable operations

### LINKED ISSUE #182403: local_map deferred inlining can expose index_put_ and force default_partition fallback
### Bug

`local_map` deferred HOP inlining can expose a mutable `aten.index_put_.default` from the local_map body in the inlined joint graph. If the graph also has any recompute metadata, `default_partition()` rejects the graph as non-functional and falls back to min-cut:

```text
Trying to unsafely apply AC to a non-functional graph with the default partitioner. Falling back to min-cut partitioner.
```

This is related to #167568, but this repro does not require a Triton mutation. It uses an ordinary ATen indexed write.

### Minimal repro

```python
from collections import Counter
from contextlib import ExitStack, contextmanager
import warnings

import torch
from torch import nn
from torch._dynamo.functional_export import dynamo_graph_capture_for_export
from torch._dynamo.variables.higher_order_ops import LocalMapWrappedHigherOrderVariable
from torch._functorch._aot_autograd.functional_utils import _is_functional_graph
from torch._functorch.aot_autograd import aot_export_joint_with_descriptors
from torch._functorch.partitioners import default_partition, has_recomputable_ops
from torch._subclasses.fake_tensor import FakeTensorMode
from torch.fx.experimental.proxy_tensor import make_fx
from torch.fx.experimental.symbolic_shapes import ShapeEnv
from torch.testing._internal.distributed.fake_pg import FakeStore
from torch.utils.checkpoint import CheckpointPolicy

from torch.distributed._tensor.experimental import local_map
from torch.distributed.tensor.placement_types import Replicate


@contextmanager
def enable_local_map_wrapping():
    from torch._higher_order_ops import local_map as local_map_module
    with LocalMapWrappedHigherOrderVariable.enable(), local_map_module.defer_inlining():
        yield


def capture_deferred_joint(model, inputs_fn):
    fake_mode = FakeTensorMode()
    fake_mode.shape_env = ShapeEnv()
    fake_mode.static_shapes = False
    with fake_mode:
        inputs = inputs_fn()

    with enable_local_map_wrapping(), torch._dynamo.utils._disable_saved_tensors_hooks_during_tracing():
        gm = dynamo_graph_capture_for_export(model)(*inputs)
        stack = ExitStack()
        try:
            return aot_export_joint_with_descriptors(stack, gm, inputs).graph_module
        finally:
            stack.close()


def inline_with_preserved_meta(gm):
    with torch.fx.traceback.preserve_node_meta():
        inputs = [n.meta["val"] for n in gm.graph.nodes if n.op == "placeholder"]
        return make_fx(torch.fx.Interpreter(gm).run)(*inputs)


def first_fw_compute_node(gm):
    for node in gm.graph.nodes:
        if node.op == "call_function" and node.meta.get("partitioner_tag") == "is_forward":
            return node
    raise AssertionError("no forward compute node found")


class M(nn.Module):
    def __init__(self, mesh):
        super().__init__()
        self.w = nn.Parameter(torch.randn(16, 16))
        self.mesh = mesh

    def forward(self, x):
        placements = (Replicate(), Replicate())

        @local_map(
            out_placements=(placements,),
            in_placements=(placements, placements),
            redistribute_inputs=True,
            in_grad_placements=None,
            device_mesh=self.mesh,
        )
        def f(x, w):
            y = x @ w
            idx = torch.argsort(x[:, 0])
            out = y.new_empty(y.shape)
            out[idx, :] = y
            return out

        return f(x, self.w).sum()


def summarize(label, gm):
    nodes = list(gm.graph.nodes)
    error, mutation_count = _is_functional_graph(gm.graph)
    mutable = [
        n
        for n in nodes
        if isinstance(n.target, torch._ops.OpOverload) and n.target._schema.is_mutable
    ]
    print(f"=== {label} ===")
    print("nodes", len(nodes))
    print("partitioner tags", Counter(n.meta.get("partitioner_tag") for n in nodes))
    print("has recompute", has_recomputable_ops(gm))
    print("mutation_count", mutation_count)
    print("functional_error", error.split("graph=")[0] if error else None)
    print("mutable targets", [(n.name, str(n.target), n.meta.get("partitioner_tag")) for n in mutable])


torch.distributed.init_process_group("fake", store=FakeStore(), rank=0, world_size=4)
mesh = torch.distributed.device_mesh.init_device_mesh("cpu", (2, 2), mesh_dim_names=("dp", "tp"))

joint = capture_deferred_joint(M(mesh), lambda: (torch.randn(8, 16),))
inlined = inline_with_preserved_meta(joint)
summarize("inlined local_map joint graph", inlined)

first_fw_compute_node(inlined).meta["recompute"] = CheckpointPolicy.MUST_RECOMPUTE
summarize("after adding one recompute tag", inlined)

with warnings.catch_warnings(record=True) as caught:
    warnings.simplefilter("always")
    default_partition(inlined, [], num_fwd_outputs=1, static_lifetime_input_indices=[])

print([str(w.message) for w in caught if "Falling back to min-cut partitioner" in str(w.message)])
torch.distributed.destroy_process_group()
```

### Actual output

```text
=== inlined local_map joint graph ===
functional_error aot_autograd expected to have an entirely functional graph, but found %index_put_ = call_function[target=torch.ops.aten.index_put_.default](...)
mutable targets [('index_put_', 'aten.index_put_.default', 'is_forward')]

=== after adding one recompute tag ===
has recompute True
functional_error aot_autograd expected to have an entirely functional graph, but found %index_put_ = call_function[target=torch.ops.aten.index_put_.default](...)

['Trying to unsafely apply AC to a non-functional graph with the default partitioner. Falling back to min-cut partitioner.']
```

### Expected behavior

The nested local_map graph should be functionalized before it is exposed to AOTAutograd partitioning. The inlined graph should contain a functional equivalent of the indexed update, not `aten.index_put_.default`, so `default_partition()` does not fall back solely because the local_map body had an indexed write into a fresh tensor.

### Impact

This shows up in larger local_map MoE workloads. For example, token unpermute code commonly has this shape:

```python
out_unpermuted = routed_output.new_empty(input_shape)
out_unpermuted[permuted_indices, :] = routed_output
```

When recompute tags are present, the remaining `index_put_` makes the default partitioner fall back to min-cut.

### Suggested fix

Ensure local_map deferred inlining functionalizes nested forward/backward GraphModules before replay/inlining, or avoid disabling/bypassing functionalization for the local_map body in the path that builds the nested joint graph. In this repro, `index_put_` should become functional `index_put` or an equivalent functional scatter/update before `default_partition()` sees the graph.

cc @jerryzh168 @bdhirsh @ezyang @chauhang @penguinwu @ydwu4 @bobrenjc93 @aorenste

### PR DESCRIPTION
**Pull request by Arindamsamanta2004**
# Fix local_map deferred inlining to functionalize mutable operations

Fixes #182403

## Problem

When `local_map` uses deferred inlining with activation checkpointing, mutable operations like `index_put_` were not functionalized before partitioning. This caused the default partitioner to reject the graph and fall back to min-cut partitioning.

**Failing scenario:**
```python
@local_map(...)
def f(x, w):
    y = x @ w
    idx = torch.argsort(x[:, 0])
    out = y.new_empty(y.shape)
    out[idx, :] = y  # index_put_ operation
    return out
```

**Before fix:**
- Error: `aot_autograd expected functional graph, but found index_put_`
- Warning: `Falling back to min-cut partitioner`
- MoE token unpermute patterns broken

## Solution

Apply post-trace functionalization: after `make_fx` traces the joint forward-backward graph, convert mutable operations (`index_put_`, `copy_`) to functional equivalents before the partitioner runs.

**Changes:**
- Added `_functionalize_fw_gm()` helper to convert mutable ops
- Called after tracing, before partitioning
- 24 lines added to `torch/_higher_order_ops/local_map.py`

## Testing

**Existing tests pass:**
- `test/higher_order_ops/test_local_map.py`: 14/14 tests pass
- `test/distributed/tensor/experimental/test_local_map.py`: 6/6 tests pass

**Issue reproduction fixed:**
- Graph is functional (no `index_put_` errors)
- Default partitioner used (no min-cut fallback)
- Verified with minimal_repro from issue

cc @bdhirsh @ezyang @ydwu4 @penguinwu @xmfan @sanketpurandare

### CODE REVIEW

File: torch/_higher_order_ops/local_map.py [hunk 1]
```diff
@@ -204,6 +204,26 @@ def __call__(self, gm: GraphModule, *args: Any, **kwargs: Any) -> Any:
 redirect_to_mode(local_map_hop, _CachedTorchDispatchMode)
 
 
+def _functionalize_fw_gm(fw_gm: GraphModule) -> GraphModule:
```
Comment (morrison-turnansky) [inline torch/_higher_order_ops/local_map.py]: This feels like we are inventing the wheel here. There is a lot of functionalization tooling already. We can either use an available tool or we need to make this function much more rigorous. Wouldn't this fail in the same type of way for other mutable ops besides index_put_ and copy_? explore available functionalization tools for this use case and if they don't work explain why to highlight a gap in the infra. @stmcgovern do you have some more context here.

**Comment by morrison-turnansky** [comment]
see comment on pr, also add a test to verify repro issue. you should be able to add a minimal test that verifies the functinalization works as opposed to the full repro.

File: torch/_higher_order_ops/local_map.py (hunk 1 above)
Comment (Arindamsamanta2004) [reply torch/_higher_order_ops/local_map.py]: Yes, current fix only covers ops from the reported issue. I will check if existing functionalization infra can be used instead of adding a custom helper and If they don't work here, I'll document it.

**Comment by Arindamsamanta2004** [comment]
Hi @morrison-turnansky , I've updated the PR to use `torch.func.functionalize` instead of manually mapping operations. Now we are re-tracing the graph under functionalization, handling mutable ops automatically.
  
About performance, I benchmarked the re-tracing overhead. It adds ~7ms at compile time in the worst case, with early exit being basically free for graphs that are already functional. Since this only runs during compilation and local_map graphs are small, I think this should be acceptable, but let me know if you see any concerns with this approach.
  
I also added tests for index_put_ and scatter_ that verify the graph is actually functional after transformation and checking that no mutable operations remain in the final graph.

**Comment by zou3519** [comment]
@claude review these changes

**Comment by Arindamsamanta2004** [comment]
I'll fix the unused `mutation_count` variable. Should I also remove the redundant mutable_ops scan in `test_functionalization_with_index_put`?  since `_is_functional_graph` already checks this.

About metadata loss from re-tracing, the tests pass through `copy_fwd_metadata_to_bw_nodes` successfully, which suggests the metadata like `seq_nr` is surviving. But I'm happy to add explicit placeholder preservation for debugging if you think it would help?

And all the 18 CI checks failed are not related to this PR. It is a flake from #163953. Once I have your feedback, I'll rebase on main to clear them.
@zou3519 @morrison-turnansky

**Review by zou3519 — COMMENTED** [review]
so https://github.com/pytorch/pytorch/pull/185043 is related and there is some discussion to be had (see my comments on that PR)

**Comment by morrison-turnansky** [comment]
posted comment at https://github.com/pytorch/pytorch/pull/185043#discussion_r3362555243 to consolidate conversation
